### PR TITLE
Automatically discover Azure Virtual Machines

### PR DESCRIPTION
**Pull request by atburke**
This PR allows Teleport to discover (but not yet enroll) Azure Virtual Machines. Implements the discovery part of #16957.

### CODE REVIEW

File: lib/srv/server/azure_watcher_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+			wantVMs: []string{"vm2", "vm4"},
+		},
+	}
+
+	for _, tc := range tests {
+		tc.matcher.Types = []string{"vm"}
+		tc.matcher.Subscriptions = []string{"sub1"}
+
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, _ := context.WithTimeout(context.Background(), 100*time.Millisecond)
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/srv/server/azure_watcher_test.go]: The timeout is just to make sure that the test doesn't hang forever if it fails.

File: lib/srv/server/watcher.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+					continue
+				}
+				log.WithError(err).Error("Failed to fetch instances")
+				continue
+			}
+			for _, inst := range instancesColl {
+				select {
+				case w.InstancesC <- inst:
+				case <-w.ctx.Done():
+				}
```
Comment (ravicious) [inline lib/srv/server/watcher.go:62]: t

File: lib/srv/server/azure_watcher.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// AzureWatcher allows callers to discover Azure virtual machines matching specified filters.
+type AzureWatcher struct {
+	// InstancesC can be used to consume newly discoverd Azure virtual machines.
+	InstancesC chan AzureInstances
+
+	fetchers      []*azureInstanceFetcher
+	fetchInterval time.Duration
+	ctx           context.Context
```
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/srv/server/azure_watcher.go]: > We should not keep the context as a part of any struct. I don't think we should fix this as a part of this PR, but we should refactor this code (Azure and AWS) and remove `ctx` from here as the next PR.

What should it look like without storing the context in a struct?

I imagine `Watcher.Run` could receive the context as an argument. Then I suppose we could add a private channel that gets closed by `Watcher.Stop`. Wherever we currently do `<-w.ctx.Done()`, we'd also need to add another select branch where we wait for that channel to get close.

OTOH, in a scenario where `Watcher.Run` receives the context as an argument, I don't think we need `Watcher.Stop` at all? I'm looking at how it's used and it looks like we wait for the parent context to get canceled and then call `Watcher.Stop` if it happens. That doesn't seem to be necessary if `Watcher.Run` already stops blocking as soon as the parent context is canceled.

**Review by ravicious — APPROVED** [review]

File: lib/srv/server/azure_watcher.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakule) [reply lib/srv/server/azure_watcher.go]: @ravicious I think that the rule of thumb is to not store context in a struct but pass it as a first argument only to the methods that need it. There is even an article on the Golang blog about it https://go.dev/blog/context-and-structs
`Close()` method is a completely different topic, and using it, especially in a multi-threaded environment, can be quite tricky. If you REALLY must use it, I'd recommend just creating a channel and closing it when calling `Close()`.

File: lib/srv/server/watcher.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// Fetcher fetches instances from a particular cloud provider.
+type Fetcher[T any] interface {
```
Comment (jakule) [inline lib/srv/server/watcher.go:28]: Do we really need generics here? I think we could just use an interface here instead of a concrete type.

**Review by jakule — APPROVED** [review]
I left a few comments. Looks good. I think that a few things can be simplified tho.

**Pull request merged**